geomesa-kafka/geomesa-kafka-utils/src/main/scala/org/locationtech/geomesa/kafka/consumer/ThreadedConsumer.scala

```diff
@@ -0,0 +1,103 @@
+import org.apache.kafka.common.errors.WakeupException
+
+import scala.util.control.NonFatal
+
+trait ThreadedConsumer extends Closeable with LazyLogging {
+
+  import scala.collection.JavaConverters._
+
+  protected val consumers: Seq[Consumer[Array[Byte], Array[Byte]]]
+  protected val topic: String
+  protected val frequency: Long
+
+  protected def closeConsumers: Boolean = true
+
+  protected def consume(record: ConsumerRecord[Array[Byte], Array[Byte]]): Unit
+
+  private val executor: ExecutorService = Executors.newFixedThreadPool(consumers.length)
+
+  private val closed = new AtomicBoolean(false)
+
+  private val commitCallback = new OffsetCommitCallback() {
+    override def onComplete(offsets: java.util.Map[TopicPartition, OffsetAndMetadata], exception: Exception): Unit = {
+      lazy val o = offsets.asScala.map { case (tp, om) => s"[${tp.topic}:${tp.partition}:${om.offset}]"}.mkString(",")
+      if (exception == null) {
+        logger.trace(s"Consumer committed offsets: $o")
+      } else {
+        logger.error(s"Consumer error committing offsets: $o : ${exception.getMessage}", exception)
+      }
+    }
+  }
+
+  def startConsumers(): Unit = {
+    consumers.foreach(c => executor.execute(new ConsumerRunnable(c, frequency)))
+    logger.debug(s"Started ${consumers.length} consumer(s) on topic $topic")
+  }
+
+  override def close(): Unit = {
+    closed.set(true)
+    consumers.foreach(_.wakeup())
+    executor.shutdownNow()
+    executor.awaitTermination(1, TimeUnit.SECONDS)
+  }
+
+  class ConsumerRunnable(val consumer: Consumer[Array[Byte], Array[Byte]], frequency: Long) extends Runnable {
+
+    private var errorCount = 0
+
+    override def run(): Unit = {
+      try {
+        while (!closed.get()) {
+          try {
+            val result = consumer.poll(frequency)
+            lazy val topics = result.partitions.asScala.map(tp => s"[${tp.topic}:${tp.partition}]").mkString(",")
+            logger.debug(s"Consumer poll received ${result.count()} records for $topics")
+            if (!result.isEmpty) {
+              val records = result.iterator()
+              while (records.hasNext) {
+                consume(records.next())
+              }
+              logger.trace(s"Consumer finished processing ${result.count()} records from topic $topics")
+              // we commit the offsets so that the next poll doesn't return the same records
+              consumer.commitAsync(commitCallback)
+              errorCount = 0 // reset error count
+            }
+          } catch {
+            case e: WakeupException => throw e
+            case e: InterruptedException => throw e
+            case NonFatal(e) =>
+              logger.warn(s"Error receiving message from topic $topic:", e)
+              errorCount += 1
+              if (errorCount < 300) { Thread.sleep(1000) } else {
+                logger.error("Shutting down due to too many errors")
+                throw new InterruptedException("Too many errors")
```

Comment:
TODO this will get caught and suppressed in the outer loop - we want it to propagate out so that it's obvious the thread died